testsuite/tests/typing-misc/let_rec_approx.ml

```diff
@@ -33,6 +33,9 @@ Error: This pattern matches values of type "string"
 let rec opt_ok_f () = opt_ok_g ~foo:A ~bar:A ()
 and opt_ok_g ?(foo : M.t option) ?(bar : M.t = M.A) () = opt_ok_f ()
 [%%expect{|
-val opt_ok_f : unit -> 'a = <fun>
-val opt_ok_g : ?foo:M.t -> ?bar:M.t -> unit -> 'a = <fun>
+Line 2, characters 35-44:
+2 | and opt_ok_g ?(foo : M.t option) ?(bar : M.t = M.A) () = opt_ok_f ()
+                                       ^^^^^^^^^
+Error: This pattern matches values of type "M.t"
+       but a pattern was expected which matches values of type "'a option"
 |}]
```

Comment:
@stedolan, you introduced that test in #12428 to illustrate a regression after the improvement to type-based disambiguation you made in #12315. None of these patches are included in flambda backend and all the related tests fail there. 

This PR includes a refactoring of `type_approx` in 7424f438630451a019f8d0e2c7cb0818f8e33630 and I was able to re-apply your improvement (155690a72f1d0f4d671691798a2967f73b6c1d69), but so far I failed to do the same with the follow-up fix (84e4d474de4acea497dabc1e8777386d6506fa03). Would you, by any chance, have some time to look at it ?

Reply:
I'll have a look at this, but I think it'll be next week before I find time.

Reply:
Thank you, meanwhile I will follow Leo's suggestion in https://github.com/ocaml/ocaml/pull/13806#discussion_r1958427530 and see if it leads me to a solution.

Reply:
This is sorted out in [7fecb8a](https://github.com/ocaml/ocaml/pull/13806/commits/7fecb8aeb93cd64d3baa3c075235f66c5ad2e2a8)